libpromises/evalfunction.c

```diff
@@ -1674,6 +1680,211 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
+    const JsonElement *e;
+    while ((e = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true)))
+    {
+        const char *key = JsonIteratorCurrentKey(&iter);
+        const char *value = JsonPrimitiveGetAsString(e);
+
+        if (0 == strcmp(key, "curl.timeout"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting timeout to %ld seconds", fp->name, IntFromString(value));
+            curl_easy_setopt(curl, CURLOPT_TIMEOUT, IntFromString(value));
+        }
+        else if (0 == strcmp(key, "curl.verbose"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting verbosity to %ld", fp->name, IntFromString(value));
+            curl_easy_setopt(curl, CURLOPT_VERBOSE, IntFromString(value));
+        }
+        else if (0 == strcmp(key, "curl.header"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting inline headers to %ld", fp->name, IntFromString(value));
+            curl_easy_setopt(curl, CURLOPT_HEADER, IntFromString(value));
+        }
+        else if (0 == strcmp(key, "curl.referer"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting referer to %s", fp->name, value);
+            curl_easy_setopt(curl, CURLOPT_REFERER, value);
+        }
+        else if (0 == strcmp(key, "curl.user-agent"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting user agent string to %s", fp->name, value);
+            curl_easy_setopt(curl, CURLOPT_USERAGENT, value);
+        }
+        else if (0 == strcmp(key, "cfengine.max_content"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting max contents to %ld", fp->name, IntFromString(value));
+            max_content = IntFromString(value);
+        }
+        else if (0 == strcmp(key, "cfengine.max_headers"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting max headers to %ld", fp->name, IntFromString(value));
+            max_headers = IntFromString(value);
+        }
+    }
+
+    JsonElement *options_headers = JsonObjectGetAsArray(options, "curl.headers");
+    iter = JsonIteratorInit(options_headers);
+    struct curl_slist *header_list = NULL;
+    while ((e = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true)))
+    {
+        header_list = curl_slist_append(header_list, JsonPrimitiveGetAsString(e));
+    }
+
+    if (NULL != header_list)
+    {
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, header_list);
+    }
+
+    JsonElement *result = JsonObjectCreate(10);
+    CURLcode res = curl_easy_perform(curl);
+    if (NULL != header_list)
+    {
+        curl_slist_free_all(header_list);
+        header_list = NULL;
+    }
+
+    long returncode = 0;
+    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &returncode);
+    JsonObjectAppendInteger(result, "returncode",  returncode);
+
+    curl_easy_cleanup(curl);
+
+    JsonObjectAppendInteger(result, "rc",  0+res);
+
+    JsonObjectAppendBool(result, "success",  res == 0);
+
+    BufferTrimToMaxLength(content, max_content);
```

Comment:
We can't control the chunking that libcurl does, but we can refuse to accept its data.  If the write function returns 0, as it will when we're over the limit, libcurl will take the hint.

Unfortunately if we limit the size of the headers, libcurl assumes it can't write the content **at all**.  I think that's not fixable on our side.